File: p024/p24.php
<?php

/**
A permutation is an ordered arrangement of objects.  For example, 3124 is one possible permutation of the
digits 1, 2, 3 and 4.  If all of the permutations are listed numerically or alphabetically, we call it
lexicographic order.  The lexicographic permutations of 0, 1, 2 are:

012	021	102	120	201	210

What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9?
**/

function factorial($n) {
	$total = 1;

	if($n <= 1) return 1;
	for($i = $n; $i > 0; $i--) {
		$total *= $i;
	}

	return $total;
}

function lexicographic_index($p) {
	$result = 0;

	for($j = 0; $j < count($p); $j++) {
		$k = 0;
		for($i = $j + 1; $i < count($p); $i++) {
			if($p[$i] < $p[$j]) {
				$k += 1;
			}
		}

		$result += $k * factorial(count($p) - $j - 1);
	}

	return $result;
}

function lexicographic_followers($p) {
	return factorial(count($p)) - lexicographic_index($p) - 1;
}

function print_perm($p) {
	for($i = 0; $i < count($p); $i++) {
		print $p[$i];
	}
}

$limit = 1000000;
$p = array(0,0,0,0,0,0,0,0,0,0);
$nums = array(0,1,2,3,4,5,6,7,8,9);

for($i = 0; $i < count($p); $i++) {

	// determine the next maximum number for the next slot in the permutation
	// go through the remaining available numbers
	// append the 
	$max = -1;
	$max_index = -1;
	for($j = 0; $j < count($nums); $j++) {
		$p[$i] = $nums[$j];
		for($k = 0; $k < count($nums); $k++) {
			if(!in_array($nums[$k], $p) && $i + $k + 1 < count($p)) {
				$p[$i + $k + 1] = $nums[$k];
			}
		}
		$t = lexicographic_index($p);

		if($t >= $limit) continue;
		if($t > $max) {
			$max = $t;
			$max_index = $j;
		}
	}

	$p[$i] = $nums[$max_index];
	array_splice($nums, $max_index, 1);
}

print_perm($p);

// answer determined by manually checking hightest to lowest numbers, the algorithm
// above does the same but automated
//print lexicographic_index(array(2,7,8,3,9,1,5,4,6,0)) . "\n";

?>

File: p024/p24.c
#include <stdio.h>
#include <stdbool.h>

typedef long long int64;

int64 factorial(int64 n) {
	int f = 1;
	if(n <= 1) return 1;
	for(int64 i = n; i > 0; i--) f *= i;
	return f;
}

int64 lex_index(int* p) {
	int64 result = 0;

	for(int64 j = 0; j < 10; j++) {
		int64 k = 0;
		for(int64 i = j + 1; i < 10; i++) {
			if(p[i] < p[j]) k++;
		}

		result += k * factorial(10 - j - 1);
	}

	return result + 1;
}

int main(int argc, char* argv[]) {

	int p[10] = { -1, -1, -1, -1, -1, -1, -1, -1 , -1, -1 };
	int nums[10] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };

	for(int i = 0; i < 10; i++) {
		int64 max = -1;
		int max_index = -1;

		for(int j = 0; j < 10; j++) {
			if(nums[j] == -1) continue;

			p[i] = nums[j];
			int z = 1;
			for(int k = 0; k < 10; k++) {
				if(nums[k] == -1) continue;

				bool exists = false;
				for(int e = 0; e <= i; e++) {
					if(p[e] == nums[k]) exists = true;
				}

				if(!exists && i + z < 10) {
					p[i + z] = nums[k];
					z++;
				}
			}
			int64 t = lex_index(p);

			if(t > 1000000) continue;
			if(t > max) {
				max = t;
				max_index = j;
			}
		}

		p[i] = nums[max_index];
		nums[max_index] = -1;
		for(int j = i + 1; j < 10; j++) p[j] = -1;
	}

	for(int i = 0; i < 10; i++) {
		printf("%i", p[i]);
	}

	//int64 i = lex_index(2783915460);
	//printf("%llu", i);

	return 0;
}

